File: soft2/lec07/python/echo-client.py
#!/usr/bin/env python

# http://omniorb.sourceforge.net/omnipy3/omniORBpy/omniORBpy002.html
import sys
from omniORB import CORBA
import CosNaming, EchoApp

# Initialise the ORB
orb = CORBA.ORB_init(sys.argv, CORBA.ORB_ID)

# Obtain a reference to the root naming context
obj         = orb.resolve_initial_references("NameService")
rootContext = obj._narrow(CosNaming.NamingContext)

if rootContext is None:
    print "Failed to narrow the root naming context"
    sys.exit(1)

# Resolve the name "test.my_context/ExampleEcho.Object"
name = [CosNaming.NameComponent("context", ""),
        CosNaming.NameComponent("Example", "")]
try:
    obj = rootContext.resolve(name)

except CosNaming.NamingContext.NotFound, ex:
    print "Name not found"
    sys.exit(1)

# Narrow the object to an Example::Echo
eo = obj._narrow(EchoApp.Echo)

if eo is None:
    print "Object reference is not an Example::EchoApp"
    sys.exit(1)

# Invoke the echoString operation
message = "Hello from Python"
result  = eo.echoString(message)

print "I said '%s'. The object said '%s'." % (message,result)

File: soft2/lec07/cpp/echo-client.cpp
// http://omniorb.sourceforge.net/omni41/omniORB/omniORB002.html#toc13
// https://xennis.org/wiki/CORBA_-_Advanced_example_with_server-client_in_Java_and_C%2B%2B#C.2B.2B_non-interactive_client
#include "echo.hh"
#include <iostream>

EchoApp::Echo_ptr service_server;
using namespace std;

/** Name is defined in the server */
#define SERVER_NAME		"Example"

int main(int argc, char ** argv)
{
  try {
    //------------------------------------------------------------------------
    // Initialize ORB object.
    //------------------------------------------------------------------------
    CORBA::ORB_ptr orb = CORBA::ORB_init(argc, argv);

    //------------------------------------------------------------------------
    // Resolve service
    //------------------------------------------------------------------------
    service_server = 0;

    try {
      
      //------------------------------------------------------------------------
      // Bind ORB object to name service object.
      // (Reference to Name service root context.)
      //------------------------------------------------------------------------
      CORBA::Object_var ns_obj = orb->resolve_initial_references("NameService");
      
      if (!CORBA::is_nil(ns_obj)) {
	//------------------------------------------------------------------------
	// Bind ORB object to name service object.
	// (Reference to Name service root context.)
	//------------------------------------------------------------------------
	CosNaming::NamingContext_ptr nc = CosNaming::NamingContext::_narrow(ns_obj);
	
	//------------------------------------------------------------------------
	// The "name text" put forth by CORBA server in name service.
	// This same name ("MyServerName") is used by the CORBA server when
	// binding to the name server (CosNaming::Name).
	//------------------------------------------------------------------------
	CosNaming::Name name;
	name.length(2);
	name[0].id = CORBA::string_dup("context");
	name[0].kind = CORBA::string_dup("");
	name[1].id = CORBA::string_dup(SERVER_NAME);
	name[1].kind = CORBA::string_dup("");

	//------------------------------------------------------------------------
	// Resolve "name text" identifier to an object reference.
	//------------------------------------------------------------------------
	CORBA::Object_ptr obj = nc->resolve(name);

	if (!CORBA::is_nil(obj)) {
	  service_server = EchoApp::Echo::_narrow(obj);
	}
      }
    } catch (CosNaming::NamingContext::NotFound &) {
      cerr << "Caught corba not found" << endl;
    } catch (CosNaming::NamingContext::InvalidName &) {
      cerr << "Caught corba invalid name" << endl;
    } catch (CosNaming::NamingContext::CannotProceed &) {
      cerr << "Caught corba cannot proceed" << endl;
    }
    
    //------------------------------------------------------------------------
    // Do stuff
    //------------------------------------------------------------------------
    if (!CORBA::is_nil(service_server)) {
      cout << "EchoApp client is running ..." << endl;
      string result = service_server->echoString("Hello from C++");
      cerr << "The object said " << result << endl;
    }
    //------------------------------------------------------------------------
    // Destroy OBR
    //------------------------------------------------------------------------
    orb->destroy();
  } catch (CORBA::UNKNOWN) {
    cerr << "Caught CORBA exception: unknown exception" << endl;
  }
}
